### PR TITLE
feature: mr2 world map

### PR DESCRIPTION
**Pull request by sush1lemon**
- [x] Implement join world map
- [x] Implement getArea | get bases
- [x] Implement leave world | recycle map room
- [x] Load correct bases on map room
- [x] Attack correct bases on map room
- [x] Save attackers and attacked base updates
- [x] Implement save bookmarks
- [x] Implement outposts
- [x] Implement outpost take over
- [x] Implement wild tribes
- [x] Fix catapult error
- [x] Implement monsters transfer
- [x] Implement repair / base update of wild tribes
- [x] Implement base migrate
- [ ] Implement creating new world
- [ ] Create base save with different tribes and level

### CODE REVIEW

File: client/scripts/com/monsters/maproom_advanced/MapRoomCell.as [hunk 1]
```diff
@@ -15,7 +15,8 @@ package com.monsters.maproom_advanced
    public class MapRoomCell extends MapRoomCell_CLIP implements IMapRoomCell
    {
        
-      
+      internal var _cell_id:int;
```
Comment (React1-X) [inline client/scripts/com/monsters/maproom_advanced/MapRoomCell.as:18]: Reminder: Remove any new client side logic, we should emulate the original server.

File: client/scripts/com/monsters/maproom_manager/MapRoomManager.as [hunk 2]
```diff
@@ -325,7 +325,7 @@ package com.monsters.maproom_manager
          }
          if(MapRoomManager.instance.isInMapRoom2 == false)
          {
-            MapRoom3ConfirmMigrationPopup.instance.Show(true);
+            // MapRoom3ConfirmMigrationPopup.instance.Show(true);
```
Comment (React1-X) [inline client/scripts/com/monsters/maproom_manager/MapRoomManager.as:328]: This seems fine, if we're going to modify the client in any way, this would be an acceptable change.

File: server/src/app.routes.ts [hunk 3]
```diff
@@ -43,24 +45,29 @@ router.post("/base/updatesaved", auth, debugDataLog("Base updated save"), update
 router.get("/api/bm/yardplanner/gettemplates", auth, debugDataLog("Get templates"), getTemplates);
 router.post("/api/bm/yardplanner/savetemplate", auth, debugDataLog("Saving template"), saveTemplate);
 
+
+router.post("/api/bm/neighbours/get", debugDataLog("Getting user neighbors"), async (ctx) => {
```
Comment (React1-X) [inline server/src/app.routes.ts:49]: Although small in size, we should move this into it's own controller, for consistency.

File: server/src/app.routes.ts [hunk 4]
```diff
@@ -43,24 +45,29 @@ router.post("/base/updatesaved", auth, debugDataLog("Base updated save"), update
...
+  ctx.status = 200;
+  ctx.body = {
+    error: 0,
+  }
+})
```
Comment (React1-X) [inline server/src/app.routes.ts:54]: Although small in size, we should move this into it's own controller, for consistency.

File: client/scripts/com/monsters/maproom_manager/MapRoomManager.as (hunk 2 above)
Comment (React1-X) [inline client/scripts/com/monsters/maproom_manager/MapRoomManager.as:328]: This seems fine, if we're going to modify the client in any way, this would be an acceptable change.

File: client/scripts/com/monsters/maproom_advanced/MapRoomCell.as (hunk 1 above)
Comment (React1-X) [inline client/scripts/com/monsters/maproom_advanced/MapRoomCell.as:18]: Reminder: Remove any new client side logic, we should emulate the original server.

File: server/src/controllers/maproom/v2/cells/homeCell.ts [hunk 5]
```diff
@@ -13,11 +14,12 @@ export const homeCell = async (ctx: Context) => {
...
    */
   return {
-    uid: save.userid, // ToDo: Why do we have a userId for both user and save table? Fix
+    cell_id: cell.cell_id,
+    uid: cell.uid, // ToDo: Why do we have a userId for both user and save table? Fix
     b: 2,
     fbid: save.fbid,
     pi: 0,
-    bid: save.homebaseid,
+    bid: save.basesaveid,
```
Comment (React1-X) [inline server/src/controllers/maproom/v2/cells/homeCell.ts:22]: Hmm... are you sure about this change? I will need some explanation. If you look in `MapRoomCell.as` we have this:

```javascript
         if(serverData.bid)
         {
            if(this._baseID != 0 && this._baseID == GLOBAL._homeBaseID)
            {
               MapRoom._homeCell = this;
            }
            else if(this.X == GLOBAL._mapHome.x && this.Y == GLOBAL._mapHome.y)
            {
               MapRoom._homeCell = this;
            }
            this._baseID = serverData.bid;
         }
```

It seems it checks if `bid` is equal to `homebaseid`, is there any reason why we have set it to be `basesaveid`?

File: server/src/controllers/maproom/v2/cells/enemyCell.ts [hunk 6]
```diff
@@ -0,0 +1,39 @@
+import { WorldMapCell } from "../../../../models/worldmapcell.model";
```
Comment (React1-X) [inline server/src/controllers/maproom/v2/cells/enemyCell.ts:1]: Is there a good reason why we created this file? Unless it is for testing purposes? I'm not sure there is such thing as an 'enemy' cell? It would just be other user's homebase, outposts and wild monster yards.

I could be wrong about my thoughts here but:

Instead of defining different cell types e.g. `enemyCell`, `wildmonsterCell`, `outpostCell` maybe we should just have a single blueprint for a cell? And then switch out the 'type', etc. Let me know what you think.

File: server/src/controllers/maproom/v2/cells/wildMonsterCell.ts [hunk 7]
```diff
@@ -1,31 +1,41 @@
...
   const randomIndex = Math.floor(Math.random() * tribes.length);
   const tribeName = tribes[randomIndex];
 
+  const save = await ORMContext.em.findOne(Save, {
+    basesaveid: cell?.base_id
+  })
+
   return {
-    uid: 0,
+    cell_id: cell?.cell_id || 0,
```
Comment (React1-X) [inline server/src/controllers/maproom/v2/cells/wildMonsterCell.ts:16]: Reminder: We should remove this, and only pass the fields the client expects (not new ones such as `cell_id`

File: server/src/mikro-orm.config.ts [hunk 8]
```diff
@@ -3,14 +3,15 @@ import { MariaDbDriver } from '@mikro-orm/mariadb';
...
+import {WorldMapCell} from "./models/worldmapcell.model";
 
 // The configuration for the ORM - Any Entities added need to be put in here, other than that probably doesn't need to be touched
 // tslint:disable-next-line: no-object-literal-type-assertion
 export default {
   type: 'mariadb',
   allowGlobalContext: false,
-  debug: true,
-  entities: [Save, User],
+  debug: false,
```
Comment (React1-X) [inline server/src/mikro-orm.config.ts:13]: Can we revert this change? We need to keep debug set to `true` for local

File: server/src/data/flags.ts [hunk 9]
```diff
@@ -2,8 +2,8 @@ import { devConfig } from "../config/DevSettings";
 
 export const flags = {
   // Platform
-  viximo: 0,
-  kongregate: 1,
+  viximo: 1,
```
Comment (React1-X) [inline server/src/data/flags.ts:5]: Is there any reasons for this change? Can we revert this?

File: client/scripts/com/monsters/maproom_advanced/MapRoomCell.as [hunk 10]
```diff
@@ -326,6 +327,7 @@ package com.monsters.maproom_advanced
          this._updated = true;
          this._processed = false;
          this._base = serverData.b;
+         this._cell_id = serverData.cell_id;
```
Comment (React1-X) [inline client/scripts/com/monsters/maproom_advanced/MapRoomCell.as:330]: Reminder: Remove any new client side logic, we should emulate the original server.

File: client/scripts/com/monsters/maproom_advanced/PopupAttackA.as [hunk 11]
```diff
@@ -142,12 +142,12 @@ package com.monsters.maproom_advanced
             GLOBAL._currentCell = this._cell;
             if(this._cell._base == 1)
             {
-               BASE.LoadBase(null,0,this._cell._baseID,"wmattack",false,EnumYardType.MAIN_YARD);
+               BASE.LoadBase(null,0,this._cell._baseID,"wmattack",false,EnumYardType.MAIN_YARD, this._cell._cell_id);
```
Comment (React1-X) [inline client/scripts/com/monsters/maproom_advanced/PopupAttackA.as:145]: Reminder: Remove any new client side logic, we should emulate the original server.

File: client/scripts/com/monsters/maproom_advanced/PopupAttackA.as [hunk 12]
```diff
@@ -142,12 +142,12 @@ package com.monsters.maproom_advanced
...
             }
             else
             {
                _loc2_ = this._cell._base == 3 ? int(EnumYardType.OUTPOST) : int(EnumYardType.MAIN_YARD);
-               BASE.LoadBase(null,0,this._cell._baseID,GLOBAL.e_BASE_MODE.ATTACK,false,_loc2_);
+               BASE.LoadBase(null,0,this._cell._baseID,GLOBAL.e_BASE_MODE.ATTACK,false,_loc2_, this._cell._cell_id);
```
Comment (React1-X) [inline client/scripts/com/monsters/maproom_advanced/PopupAttackA.as:150]: Reminder: Remove any new client side logic, we should emulate the original server.

File: client/scripts/com/monsters/maproom_advanced/PopupInfoEnemy.as [hunk 13]
```diff
@@ -764,18 +764,18 @@ package com.monsters.maproom_advanced
          GLOBAL._currentCell = this._cell;
          if(this._cell._base == 1)
          {
-            BASE.LoadBase(null,0,this._cell._baseID,GLOBAL.e_BASE_MODE.WMVIEW,false,EnumYardType.MAIN_YARD);
+            BASE.LoadBase(null,0,this._cell._baseID,GLOBAL.e_BASE_MODE.WMVIEW,false,EnumYardType.MAIN_YARD, this._cell._cell_id);
```
Comment (React1-X) [inline client/scripts/com/monsters/maproom_advanced/PopupInfoEnemy.as:767]: Reminder: Remove any new client side logic, we should emulate the original server.

File: client/scripts/com/monsters/maproom_advanced/PopupInfoEnemy.as [hunk 14]
```diff
@@ -764,18 +764,18 @@ package com.monsters.maproom_advanced
...
          }
          else
          {
             _loc1_ = this._cell._base == 3 ? int(EnumYardType.OUTPOST) : int(EnumYardType.MAIN_YARD);
             if(this._cell._friend)
             {
-               BASE.LoadBase(null,0,this._cell._baseID,GLOBAL.e_BASE_MODE.HELP,false,_loc1_);
+               BASE.LoadBase(null,0,this._cell._baseID,GLOBAL.e_BASE_MODE.HELP,false,_loc1_, this._cell._cell_id);
```
Comment (React1-X) [inline client/scripts/com/monsters/maproom_advanced/PopupInfoEnemy.as:774]: Reminder: Remove any new client side logic, we should emulate the original server.

File: client/scripts/com/monsters/maproom_advanced/PopupInfoEnemy.as [hunk 15]
```diff
@@ -764,18 +764,18 @@ package com.monsters.maproom_advanced
...
             }
             else
             {
-               BASE.LoadBase(null,0,this._cell._baseID,GLOBAL.e_BASE_MODE.VIEW,false,_loc1_);
+               BASE.LoadBase(null,0,this._cell._baseID,GLOBAL.e_BASE_MODE.VIEW,false,_loc1_, this._cell._cell_id);
```
Comment (React1-X) [inline client/scripts/com/monsters/maproom_advanced/PopupInfoEnemy.as:778]: Reminder: Remove any new client side logic, we should emulate the original server.

File: server/src/controllers/baseLoad.ts [hunk 16]
```diff
@@ -9,22 +9,50 @@ import { getDefaultBaseData } from "../data/getDefaultBaseData";
...
   await ORMContext.em.populate(user, ["save"]);
 
+
   let save = user.save;
+
+
+
+  if (requestBody.cellid && requestBody.baseid) {
+    const cell = await ORMContext.em.findOne(WorldMapCell, {
+      cell_id: parseInt(requestBody.cellid),
```
Comment (React1-X) [inline server/src/controllers/baseLoad.ts:38]: Reminder: Client doesn't expect this `cell_id`

File: server/src/controllers/baseLoad.ts [hunk 17]
```diff
@@ -9,22 +9,50 @@ import { getDefaultBaseData } from "../data/getDefaultBaseData";
...
+    })
+    if (cell) {
+      if (cell.base_type) {
+        save = await ORMContext.em.findOne(Save, {
+          basesaveid: cell.base_id,
+        })
+      }
+    }
+  }
+
```
Comment (React1-X) [inline server/src/controllers/baseLoad.ts:48]: Redundant nesting for if conditions here.

```javascript 
if (cell && cell.base_type)
```

File: server/src/controllers/maproom/v2/getArea.ts [hunk 18]
```diff
@@ -11,43 +15,85 @@ interface Cell {
...
 
-      if (x === 2 && y === 1) {
-        cells[x][y] = await outpostCell(ctx);
-      }
+      cells[x][y] = await wildMonsterCell();
+      //
+      // if (x === 2 && y === 1) {
+      //   cells[x][y] = await outpostCell(ctx);
+      // }
+      // cells[x][y] = null;
```
Comment (React1-X) [inline server/src/controllers/maproom/v2/getArea.ts:88]: I will need a general overview here of what you are trying to achieve with this new code added, and the benefits of this approach.

**Comment by React1-X** [comment]
TLDR: Good work! Thanks a lot. There is quite a lot here, so understandably it will take some time for myself and @BigmanGHark to look through it, see if it's the best approach, improve some things, etc. I've left some comments, some are questions, others are improvements that can be made. @sush1lemon

File: server/src/controllers/maproom/v2/cells/homeCell.ts (hunk 5 above)
Comment (sush1lemon) [reply server/src/controllers/maproom/v2/cells/homeCell.ts:22]: I reverted it back to homebaseid

File: server/src/controllers/maproom/v2/cells/enemyCell.ts (hunk 6 above)
Comment (sush1lemon) [reply server/src/controllers/maproom/v2/cells/enemyCell.ts:1]: Yes, it is just for testing. and yes we should use only one

File: server/src/controllers/baseLoad.ts (hunk 17 above)
Comment (sush1lemon) [reply server/src/controllers/baseLoad.ts:48]: this is actually incomplete, it should really check for the base_type because our plan is to use different table for the save, tribes, outposts right?

File: server/src/data/flags.ts (hunk 9 above)
Comment (sush1lemon) [reply server/src/data/flags.ts:5]: I think set this because there's a checking on client if the platform is viximo and it will load the pictures properly

https://github.com/bym-refitted/backyard-monsters-refitted/blob/d08ffdb72e4afda14ee69944a17ea05bd8b06944/client/scripts/com/monsters/maproom_advanced/PopupInfoEnemy.as#L497

File: server/src/controllers/baseLoad.ts (hunk 17 above)
Comment (React1-X) [reply server/src/controllers/baseLoad.ts:48]: Yeah exactly

File: server/src/controllers/maproom/v2/cells/homeCell.ts [hunk 19]
```diff
@@ -1,40 +1,68 @@
...
+      fr: 0,
+      on: online, // ToDo
+      p: save.protected,
+      r: save.resources,
+      m: save.monsters,
+      l: 1, // ToDo
+      d: save.damage > 90,
+      lo: save.locked,
+      dm: save.damage,
+      pic_square: `https://api.dicebear.com/7.x/adventurer/png?backgroundType=solid&backgroundColor=b6e3f4,c0aede,d1d4f9&seed=${user.username}`,
```
Comment (BigmanGHark) [inline server/src/controllers/maproom/v2/cells/homeCell.ts:51]: This is a pretty cool change thank you. Just thinking will we have CORS errors with this on prod. We can investigate later though

Comment (sush1lemon) [reply server/src/controllers/maproom/v2/cells/homeCell.ts:55]: [dicebear](https://www.dicebear.com/introduction/) allow all origin so i think it wouldn't be problem

image

**Review by React1-X — APPROVED** [review]
